openmeter/ent/schema/billing.go

```diff
@@ -0,0 +1,418 @@
+			}),
+	}
+}
+
+func (BillingInvoiceItem) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "invoice_id"),
+		index.Fields("namespace", "customer_id"),
+	}
+}
+
+func (BillingInvoiceItem) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("billing_invoice", BillingInvoice.Type).
+			Ref("billing_invoice_items").
+			Field("invoice_id").
+			Unique(),
+		// TODO[dependency]: Customer edge, as soon as customer entities are ready
+
+	}
+}
+
+type BillingInvoice struct {
+	ent.Schema
+}
+
+func (BillingInvoice) Mixin() []ent.Mixin {
+	return []ent.Mixin{
+		entutils.IDMixin{},
+		entutils.NamespaceMixin{},
+		entutils.TimeMixin{},
+		entutils.MetadataAnnotationsMixin{},
+	}
+}
+
+func (BillingInvoice) Fields() []ent.Field {
+	return []ent.Field{
+		field.String("key").
+			NotEmpty().
+			Immutable(),
+		field.String("customer_id").
+			NotEmpty().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}).
+			Immutable(),
+		field.String("billing_profile_id").
+			NotEmpty().
+			Immutable().
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}),
+		field.Time("voided_at").
+			Optional(),
+		field.String("currency").
+			NotEmpty().
+			Immutable().
+			SchemaType(map[string]string{
+				"postgres": "varchar(3)",
+			}),
+		field.Other("total_amount", alpacadecimal.Decimal{}).
+			SchemaType(map[string]string{
+				"postgres": "numeric",
+			}),
+		field.Time("due_date"),
+		field.Enum("status").
+			GoType(invoice.InvoiceStatus("")),
+
+		field.String("provider_config").
+			GoType(provider.Configuration{}).
+			ValueScanner(ProviderConfigValueScanner).
+			SchemaType(map[string]string{
+				"postgres": "jsonb",
+			}),
+
+		field.String("workflow_config_id").
+			SchemaType(map[string]string{
+				"postgres": "char(26)",
+			}),
+		field.String("provider_reference").
+			GoType(provider.Reference{}).
+			ValueScanner(ProviderReferenceValueScanner).
+			SchemaType(map[string]string{
+				"postgres": "jsonb",
+			}),
+
+		field.Time("period_start"),
+		field.Time("period_end"),
+	}
+}
+
+func (BillingInvoice) Indexes() []ent.Index {
+	return []ent.Index{
+		index.Fields("namespace", "id"),
+		index.Fields("namespace", "key"),
+		index.Fields("namespace", "customer_id"),
+		index.Fields("namespace", "due_date"),
+		index.Fields("namespace", "status"),
+	}
+}
+
+func (BillingInvoice) Edges() []ent.Edge {
+	return []ent.Edge{
+		edge.From("billing_profile", BillingProfile.Type).
+			Ref("billing_invoices").
+			Field("billing_profile_id").
+			Required().
+			Unique().
+			Immutable(), // Billing profile changes are forbidden => invoice must be voided in this case
+		edge.From("billing_workflow_config", BillingWorkflowConfig.Type).
+			Ref("billing_invoices").
+			Field("workflow_config_id").
+			Unique().
+			Required(),
+		edge.To("billing_invoice_items", BillingInvoiceItem.Type),
+	}
+}
+
+type providerReferenceSerde[T any] struct {
```

Comment:
Is this a generic helper?

Reply:
This is used by the parser to extract the jsonb field based on the type field. I wouldn't abstract it further as the value scanner needs to be custom either ways.